common/lib/plugins/bluegreen/blue_green_status_provider.ts

```diff
@@ -0,0 +1,939 @@
+  }
+
+  protected isBlueGreenDialect(dialect: any): dialect is BlueGreenDialect {
+    return dialect;
+  }
+
+  protected async prepareStatus(role: BlueGreenRole, interimStatus: BlueGreenInterimStatus): Promise<void> {
+    // Detect changes
+    const contextHash: number = this.getContextHash();
+
+    const storedStatus = this.interimStatuses[role.value];
+
+    if (_.isEqual(storedStatus, interimStatus) && this.lastContextHash === contextHash) {
+      // no changes detected
+      logger.debug(`no changes detected for role: ${role.name}`);
+      return;
+    }
+
+    // There are some changes detected. Let's update summary status.
+    logger.debug(Messages.get("Bgd.interimStatus", this.bgdId, role.name, interimStatus.toString()));
+
+    this.updatePhase(role, interimStatus);
+
+    // Store interimStatus and corresponding hash
+    this.interimStatuses[role.value] = interimStatus;
+    this.lastContextHash = contextHash;
+
+    // Update map of IP addresses.
+    for (const [host, ip] of interimStatus.startIpAddressesByHostMap) {
+      this.hostIpAddresses.set(host, ip);
+    }
+
+    // Update roleByHost based on provided host names.
+    interimStatus.hostNames.forEach((x) => this.roleByHost.set(x.toLowerCase(), role));
+
+    this.updateCorrespondingHosts();
+    this.updateSummaryStatus(role, interimStatus);
+    await this.updateMonitors();
+    this.updateStatusCache();
+    this.logCurrentContext();
+
+    // Log final switchover results.
+    this.logSwitchoverFinalSummary();
+
+    this.resetContextWhenCompleted();
+  }
+
+  protected updatePhase(role: BlueGreenRole, interimStatus: BlueGreenInterimStatus): void {
+    const latestInterimPhase: BlueGreenPhase = !this.interimStatuses[role.value]
+      ? BlueGreenPhase.NOT_CREATED
+      : this.interimStatuses[role.value].blueGreenPhase;
+
+    if (latestInterimPhase && interimStatus.blueGreenPhase && interimStatus.blueGreenPhase.phase < latestInterimPhase.phase) {
+      this.rollback = true;
+      logger.debug(Messages.get("Bgd.rollback", this.bgdId));
+    }
+
+    if (!interimStatus.blueGreenPhase) {
+      return;
+    }
+
+    // Do not allow status moves backward (unless it's rollback).
+    // That could be caused by updating blue/green hosts delays.
+    if (!this.rollback) {
+      if (interimStatus.blueGreenPhase.phase >= this.latestStatusPhase.phase) {
+        this.latestStatusPhase = interimStatus.blueGreenPhase;
+      }
+    } else {
+      if (interimStatus.blueGreenPhase.phase < this.latestStatusPhase.phase) {
+        this.latestStatusPhase = interimStatus.blueGreenPhase;
+      }
+    }
+  }
+
+  protected updateStatusCache(): void {
+    this.pluginService.setStatus<BlueGreenStatus>(BlueGreenStatus, this.summaryStatus, this.bgdId);
+    this.storePhaseTime(this.summaryStatus.currentPhase);
+  }
+
+  protected updateCorrespondingHosts(): void {
+    this.correspondingHosts.clear();
+
+    if (
+      this.interimStatuses[BlueGreenRole.SOURCE.value] &&
```

Comment:
It looks like there are still some fixes from the JDBC bg fix pr that need to be ported over, here and in some other spots. Let me know if you have any questions about them
https://github.com/aws/aws-advanced-jdbc-wrapper/pull/1441